interfaces/builtin/bluez.go

```diff
@@ -0,0 +1,244 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"github.com/ubuntu-core/snappy/interfaces"
+)
+
+var bluezPermanentSlotAppArmor = []byte(`
+# Description: Allow operating as the bluez service. Reserved because this
+#  gives privileged access to the system.
+# Usage: reserved
+
+  network bluetooth,
+
+  capability net_admin,
+  capability net_bind_service,
+
+  # File accesses
+  /sys/bus/usb/drivers/btusb/     r,
+  /sys/bus/usb/drivers/btusb/**   r,
+  /sys/class/bluetooth/           r,
+  /sys/devices/**/bluetooth/      rw,
+  /sys/devices/**/bluetooth/**    rw,
+  /sys/devices/**/id/chassis_type r,
+
+  # TODO: use snappy hardware assignment for this once LP: #1498917 is fixed
+  /dev/rfkill rw,
+
+  # DBus accesses
+  #include <abstractions/dbus-strict>
+  dbus (send)
+     bus=system
+     path=/org/freedesktop/DBus
+     interface=org.freedesktop.DBus
+     member={Request,Release}Name
+     peer=(name=org.freedesktop.DBus),
+
+  dbus (send)
+    bus=system
+    path=/org/freedesktop/*
+    interface=org.freedesktop.DBus.Properties
+    peer=(label=unconfined),
+
+  # Allow binding the service to the requested connection name
+  dbus (bind)
+      bus=system
+      name="org.bluez",
+
+  # Allow traffic to/from our path and interface with any method
+  dbus (receive, send)
+      bus=system
+      path=/org/bluez{,/**}
+      interface=org.bluez.*,
+
+  # Allow traffic to/from org.freedesktop.DBus for bluez service
+  dbus (receive, send)
+      bus=system
+      path=/
+      interface=org.freedesktop.DBus.**,
+  dbus (receive, send)
+      bus=system
+      path=/org/bluez{,/**}
+      interface=org.freedesktop.DBus.**,
+`)
+
+var bluezConnectedPlugAppArmor = []byte(`
+# Description: Allow using bluez service. Reserved because this gives
+#  privileged access to the bluez service.
+# Usage: reserved
+
+#include <abstractions/dbus-strict>
+
+# Allow all access to bluez service
+dbus (receive, send)
+    bus=system
+    peer=(label=bluez5_bluez_*),
```

Comment:
I just noticed this label is not going to match correctly for two reasons:

1. bluez5_bluez_* is not how the profile name is formed any more, it should be snap.bluez5.bluez.*
2. this label should match the label of the slot side. Ie, perhaps there is bluez5 bluez-foo and bluez-bar in the store that all offer the bluez slot to clients. snappy should dynamically create this plug side snippet to match the label of the slot side that the client is connecting to. Ie, if 'bluez-client-snap' is connected to 'bluez-foo', then the interface code should create a rule like this on the plugs side: dbus (receive, send) bus=system peer=(label=snap.bluez-foo.service.*)

Interestingly, the slot side could have reciprocal rules to only allow the connected clients as well.